Правки после ревью + окончание верстки 1 спринта

Exclusive048 opened the pull request.

Commit:
Правки после ревью + окончание верстки 1 спринта

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, так как проект не запускается. Скрин ошибки https://disk.yandex.ru/i/vyU9_PyA2pKrmA

Commit:
Fix error

Exclusive048 commented:
> Здравствуйте. (**Нужно развернуть общий комментарий ↓**)
> 
> К сожалению, работу придется отклонить без потери итерации, так как проект не запускается. Скрин ошибки https://disk.yandex.ru/i/vyU9_PyA2pKrmA

Добрый вечер! 
Не понимаю как так произошло. Перед отправкой, естественно, всё проверял. 
Сейчас установил заново библиотеку. Увидел что проставились зависимости. 
Прошу еще раз проверить
![image](https://github.com/user-attachments/assets/a96f63b1-80c4-42b0-ad86-268c3def0471)

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,68 @@
+        setRequesState({ ...stateRequest, hasError: false, isLoading: true });
+        let response = await fetch(ingredientsURL);
+        if (response.ok) {
+            let data = await response.json();
```
Если переменная объявлена через `let`, её значение должно быть напрямую перезаписано.
Если переменная **не переназначается**, то она объявляется через `const`.


Все переменные нужно объявлять по возможности через `const`.
В больших функциях бывает так, что Вы можете не заметить, если переменная переназначится, и это может оказаться причиной багов. Особенно в замыканиях. 
`const` дает Вам уверенность в том, что Вы всегда будете видеть одно и то же значение.

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,68 @@
+        else {
+            setRequesState({ ...stateRequest, hasError: true, isLoading: false });
+            alert("Ошибка HTTP: " + response.status);
+        }
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,68 @@
+      
+      useEffect(() => {
+        getIngredients();
+      }, []);
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -0,0 +1,68 @@
+            <AppHeader isLoading={stateRequest.isLoading} hasError={stateRequest.hasError}/>
+
+            {  !successed ? (<p>Info loading</p>) : (
+            <section className={appStyles.row}>
```
Вместо `section` нужен тег `main` по семантике тут

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,71 @@
+
+    return (
+      <>
+        <OrderDetails show={show} identifierOrder={"034536"} onCloseModal={() => setShow(false)}/>
```
Не нужно передавать в пропсы `Modal` показатель открытости `show`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.


```jsx

      {show && (
        <OrderDetails
          identifierOrder={"034536"}
          onCloseModal={() => setShow(false)}
        />
      )}
```

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,71 @@
+
+
+const BurgerConstructor = (props) => {
+    const [show, setShow] = useState(false)
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

Exclusive048 replied:
Обещаю применить это в дальнейшем. Возможно к сдаче работы по 2 спринту

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,71 @@
+
+
+  
+  const ingredient = PropTypes.shape({
+    _id: PropTypes.string.isRequired,
+    name: PropTypes.string.isRequired,
+    type: PropTypes.string.isRequired,
+    proteins: PropTypes.number.isRequired,
+    fat: PropTypes.number.isRequired,
+    carbohydrates: PropTypes.number.isRequired,
+    calories: PropTypes.number.isRequired,
+    price: PropTypes.number.isRequired,
+    image: PropTypes.string.isRequired,
+    image_mobile: PropTypes.string.isRequired,
+    image_large: PropTypes.string.isRequired,
+    __v: PropTypes.number.isRequired,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,115 @@
+  type: PropTypes.string.isRequired,
+});
+
+const data = PropTypes.shape({
+  _id: PropTypes.string.isRequired,
+  name: PropTypes.string.isRequired,
+  type: PropTypes.string.isRequired,
+  proteins: PropTypes.number.isRequired,
+  fat: PropTypes.number.isRequired,
+  carbohydrates: PropTypes.number.isRequired,
+  calories: PropTypes.number.isRequired,
+  price: PropTypes.number.isRequired,
+  image: PropTypes.string.isRequired,
+  image_mobile: PropTypes.string.isRequired,
+  image_large: PropTypes.string.isRequired,
+  __v: PropTypes.number.isRequired,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,115 @@
+        show={ingredientState.show}
+      >
+        <IngredientDetails
+          ingredientName={ingredientState.ingredientName}
+          ingredientProtein={ingredientState.ingredientProtein}
+          ingredientFat={ingredientState.ingredientFat}
+          ingredeintCarbohydrates={ingredientState.ingredeintCarbohydrates}
+          ingredientCalories={ingredientState.ingredientCalories}
+          ingredientImg={ingredientState.ingredientImg}
```
**МОЖНО ЛУЧШЕ**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента

gennady-bars commented on src/components/burger-ingredients-elem/burger-intgredients-elem.jsx:
```diff
@@ -0,0 +1,30 @@
+const BurgerIngredientsElem = (props) => {
+  return (
+    <div className={`${elemStyles.elem} ml-4 mr-5 mb-10 mt-6`} onClick={props.onClick}>
+      <img className={`${elemStyles.elem} ml-4 mr-5`} alt="previewImage" src={props.imageSrc}/>
```
`alt="previewImage" `

Каждой картинке (тегу `img`) нужен информативный атрибут `alt` (например: имя картинки вставлять), так как при ошибке загрузки картинки на сайте подсветится **название** этой картинки (которое в `alt`). Еще это нужно для **плохо видящих** и `незрячих` людей - для них есть специальные средства, которые распознают текст в `alt` и проговаривают его. Подробнее тут https://habr.com/ru/post/341810/

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,44 @@
+const IngredientDetails = (props) =>{
+    return (
+        <div className={ingredientDetailsStyle.mainContainer}>
+            <img alt="previewImage" src={props.ingredientImg}/>
```
* Каждой картинке (тегу `img`) нужен информативный атрибут `alt`

gennady-bars commented on src/components/modal/modal.module.css:
```diff
@@ -0,0 +1,29 @@
+    width: 720px;
+    cursor: auto;
+    z-index: 5;
+    position: absolute;
```
`    position: absolute;
`
Модальное окно должно быть всегда ровно посередине экрана. Для этого нужно указать ему и оверлею в `css` позицию `fixed`.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

но есть некоторые недочеты:



* В консоли ошибки https://disk.yandex.ru/i/gqgMH5mEZYrKKQ 
* В попапе ингредиента должна быть надпись  https://disk.yandex.ru/i/Hp5SslUrwNe-Ww  
* В верхней булке нужно добавлять текст `верх`, а в нижней - `низ`. Посмотрите в фигме, пожалуйста
* Верхняя и нижняя булки должны быть всегда **одинаковыми** по заданию и у них должна быть иконка `Замочек`, так как их нельзя будет удалять (только заменять перетаскиванием)
* Лишние файлы нужно удалить из проекта https://disk.yandex.ru/i/_GGmKesMkR3gIQ
* Если переменная **не переназначается**, то она объявляется через `const`.
* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.
* Не нужно передавать в пропсы `Modal` показатель открытости `show`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.
* Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `IngredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно
* Каждой картинке (тегу `img`) нужен информативный атрибут `alt` 
* Модальное окно должно быть всегда ровно посередине экрана. Для этого нужно указать ему и оверлею в `css` позицию `fixed`.




**МОЖНО ЛУЧШЕ**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента
* Можно использовать **самозакрывающиеся** теги, если нет `children` (нет ничего между тегов `><`).

Commit:
Фикс после ревью:
- Удаление ненужных файлов
- Вынос PropTypes в отдельный файл
- Исправление ошибок в консоли
- Добавление текста для булок
- Корректный инфо атрибут alt
- Исправление расположение модального окна
...

Commit:
Добавлена надпись "Детали ингредиента"

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,55 @@
+
+    return (
+      <>
+        <OrderDetails identifierOrder={"034536"} onCloseModal={() => setShow(false)}/>
```
* Кое-что не исправили https://github.com/Exclusive048/react-burger/pull/2#discussion_r1831152860 

Не нужно передавать в пропсы `Modal` показатель открытости `show`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.


```jsx

      {show && (
        <OrderDetails
          identifierOrder={"034536"}
          onCloseModal={() => setShow(false)}
        />
      )}
```

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,101 @@
+
+   return (
+    <>
+    { <Modal 
+        onCloseModal={() => setIngredientState((prevState) => ({...prevState, show: false, }))}
+        show={ingredientState.show}
```
Не нужно передавать в пропсы `Modal` показатель открытости `show`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.


```jsx

  {ingredientState.show && (
      <Modal 
        onCloseModal={() => setIngredientState((prevState) => ({...prevState, show: false, }))}
      >
        <IngredientDetails
          ingredientName={ingredientState.ingredientName}
          ingredientProtein={ingredientState.ingredientProtein}
          ingredientFat={ingredientState.ingredientFat}
          ingredeintCarbohydrates={ingredientState.ingredeintCarbohydrates}
          ingredientCalories={ingredientState.ingredientCalories}
          ingredientImg={ingredientState.ingredientImg}
        />
      </Modal>
      )}
```

gennady-bars reviewed (commented):
Уже почти все отлично, но:


* Попап с номером заказа не открывается при клике на кнопку
* Кое-что не исправили https://github.com/Exclusive048/react-burger/pull/2#discussion_r1831152860

Commit:
Отображение модального окна с заказом.

Exclusive048 commented:
У меня не получается реализовать открытие модального окна с номером заказа без явной передачи show в пропсы. 
Я бы попросил Вас дать добро на мердж, чтобы я смог 2 спринт завтра отдать на ревью. 
А к этому обязательно вернусь в процессе обучения. 
Заранее спасибо

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,68 @@
+  }, [props.show]);
+
+
+    if(!props.show){
```
`props.show` не нужно использовать внутри `Modal` теперь. Это и не дает отобразиться попапу. Сам попап теперь появляется целиком снаружи, а не внутри

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,68 @@
+      return () => {
+          window.removeEventListener("keydown", close);
+      };
+  }, [props.show]);
```
`props.show` не нужно использовать внутри `Modal` теперь. Это и не дает отобразиться попапу. Сам попап теперь появляется целиком снаружи, а не внутри

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,68 @@
+          }
+      };
+
+      if (props.show) {
```
`props.show` не нужно использовать внутри `Modal` теперь. Это и не дает отобразиться попапу. Сам попап теперь появляется целиком снаружи, а не внутри

gennady-bars reviewed (commented):
Кое-что нужно подправить:

* Не открывается попап ингредиента. И ошибка в консоли https://disk.yandex.ru/i/W76WDzKxP_vvGg 
* `props.show` не нужно использовать внутри `Modal` теперь. Это и не дает отобразиться попапу. Сам попап теперь появляется целиком снаружи, а не внутри

Commit:
Финальное исправление модального компонента

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
